weather_mv/loader_pipeline/ee.py

```diff
@@ -486,21 +487,46 @@ def convert_to_asset(self, queue: Queue, uri: str):
                 channel_names = []
                 file_name = f'{asset_name}.csv'
 
-                df = xr.Dataset.to_dataframe(ds)
-                df = df.reset_index()
-                # NULL and NaN create data-type mismatch issue in ee therefore replacing all of them.
-                # fillna fills in NaNs, NULLs, and NaTs but we have to exclude NaTs.
-                non_nat = df.select_dtypes(exclude=['datetime', 'timedelta', 'datetimetz'])
-                df[non_nat.columns] = non_nat.fillna(-9999)
+                shape = math.prod(list(ds.dims.values()))
+                # Names of dimesions, coordinates and data variables.
+                dims = list(ds.dims)
+                coords = [c for c in list(ds.coords) if c not in dims]
+                vars = list(ds.data_vars)
+                header = dims + coords + vars
 
-                # Copy in-memory dataframe to gcs.
+                # Data of dimesions, coordinates and data variables.
+                dims_data = [ds[dim].data for dim in dims]
+                coords_data = [np.full((shape,), ds[coord].data) for coord in coords]
+                vars_data = [ds[var].data.flatten() for var in vars]
+                data = coords_data + vars_data
+
+                dims_shape = [len(ds[dim].data) for dim in dims]
+
+                def get_dims_data(index: int) -> t.List[t.Any]:
+                    """Returns dimensions for the given flattened index."""
+                    return [
+                        dim[int(index / math.prod(dims_shape[i+1:])) % len(dim)] for (i, dim) in enumerate(dims_data)
+                    ]
+
+                # Copy CSV to gcs.
                 target_path = os.path.join(self.asset_location, file_name)
-                with tempfile.NamedTemporaryFile() as tmp_df:
-                    df.to_csv(tmp_df.name, index=False)
-                    tmp_df.flush()
-                    tmp_df.seek(0)
+                with tempfile.NamedTemporaryFile() as temp:
+                    with open(temp.name, 'a') as f:
+                        f.write(",".join(header) + "\n")
+                        # Write rows in batches.
+                        for i in range(0, shape, ROWS_PER_WRITE):
+                            f.write(
+                                "\n".join(
+                                    [
+                                        ",".join(map(str, i))
+                                        for i in zip(*[[*get_dims_data(i), *d[i:i + ROWS_PER_WRITE]] for d in data])
+                                    ]
+                                ).replace("NULL", "-9999")
+                                + "\n"
+                            )
```

Comment:
Let's use Python's native CSV writing library, I believe it will be less error prone. 

https://docs.python.org/3/library/csv.html

Reply:
Sounds good 👍